### DISCUSSION TITLE
Reading a compound type dataset

### DISCUSSION THREAD

**Discussion by AchilleBailly** [aldanor/hdf5-rust#217] [Questions] [answered]
I have been trying for quite some hours to read the data of a HDF5 file in Rust (in which I am a beginner).
The file I am trying to read has 2 datasets. This is the output of `h5dump` on the file:
```
DATASET "metadata" {
   DATATYPE  H5T_COMPOUND {
      H5T_ARRAY { [16] H5T_STD_U8LE } "plaintext";
      H5T_ARRAY { [16] H5T_STD_U8LE } "ciphertext";
      H5T_ARRAY { [16] H5T_STD_U8LE } "key";
      H5T_ARRAY { [16] H5T_STD_U8LE } "masks";
   }
   DATASPACE  SIMPLE { ( 60000 ) / ( 60000 ) }
}
DATASET "traces" {
   DATATYPE  H5T_STD_I8LE
   DATASPACE  SIMPLE { ( 60000, 100000 ) / ( 60000, 100000 ) }
}
```

I can read the "traces" dataset just fine but when it comes to the "metadata" dataset, I have only found a way to obtain weird data full of zeros.
I found that the corresponding type to the `CompoundType` seems to be `([u8; 16], [u8; 16], [u8; 16], [u8; 16])`. Indeed, when printing the output of `type_descriptor()` for both, they are the exact same (apart from the names).

This is the output I should get (first element):
```
([  6,  65, 195,  66,  25, 182, 248,  19, 159, 132,  97,   6,  46, 137,  42,  16], 
[ 95,  75, 255, 146, 229,  35,  24,  50,  16, 107, 238,  33,  93,  56, 106, 178], 
[ 77, 251, 224, 242, 114,  33, 254,  16, 167, 141,  74, 220, 142,  73,   4, 105], 
[  7,  84,  71, 171, 129, 171,  90, 193, 236, 131,  21, 112, 185,  95, 212,  55])
```
And what I get from Rust with the following code:
```
let file = hdf5::File::open(filename).expect("Could not open file.");
type metadata_type = ([u8; 16], [u8; 16], [u8; 16], [u8; 16]);
let t: ndarray::Array0<metadata_type> = file
            .dataset("metadata")?
            .read_slice(s![0])?;
println!("datasets: {:?}", t);
```
```
[240, 137, 17, 111, 69, 86, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 
[255, 255, 255, 255, 255, 255, 255, 255, 1, 0, 0, 0, 0, 0, 0, 0], 
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

I will gladly take any idea on how to handle this! Thank you!

**Comment by mulimoen** [comment, marked as answer]
I don't have a computer right now, could you try reading the data as this type?
```rust
#[repr(C)]
#[derive(H5Type)]
struct metadata {
    plaintext: [u8; 16]
    # the other fields
}
```

**Reply by AchilleBailly** [reply]
Thank you very much ! 
It was the first thing I tried but didn't know about the decorators (if that's what they are called) !

**Reply by mulimoen** [reply]
Think it is called a derive attribute. Did the new type result in legible data?
